doc/sphinx/language/core/primitive.rst

```diff
@@ -150,7 +150,17 @@ extraction. Instead, it has to be provided by the user (if they want to compile
 or execute the extracted code). For instance, an implementation of this module
 can be taken from the kernel of Coq (see ``kernel/parray.ml``).
 
-Primitive arrays expose a functional interface, but they are internally
-implemented using a persistent data structure :cite:`ConchonFilliatre07wml`.
-Update and access to an element in the most recent copy of an array are
-constant time operations.
+Coq's primitive arrays are persistent data structures. Semantically, a set operation
+``t.[i <- a]`` represents a new array that has the same values as ``t``, except
+at position ``i`` where its value is ``a``. The array ``t`` still exists, can
+still be used and its values were not modified. Operationally, the implementation
+of Coq's primitive arrays is optimized so that the new array ``t.[i <- a]`` does not
+copy the whole ``t``. The details are in section 2.3 of :cite:`ConchonFilliatre07wml`.
+In short, the implementation keeps one version of ``t`` as an OCaml native array and
+other versions as lists of modifications to ``t``. Accesses to the native array
+version are constant time operations. Accesses to versions where all the cells of
+the array are modified have O(n) access time, the same as a list, which defeats
+the purpose of Coq's primitive arrays. The version that is kept as the native array
```

Comment:
I would phrase this more positively, e.g.:
```suggestion
version are constant time operations. However, accesses to versions where all the cells of
the array are modified have O(n) access time, the same as a list. The version that is kept as the native array
```

Reply:
The documentation is not meant to make people happy. It explains what benefits people get when they use the tools appropriately, and it also warns about the dangers when they use the tools badly. In an algorithm with many sets and gets, it is not clear that primitive arrays are faster than lists.

Reply:
It's not about happiness, but clarity. The manual to give time complexities and explains when they are advantageous. If you need to say "defeat the purpose", it means you've said something inaccurate which you are now correcting, which is less clear.